### PR TITLE
Stack execution

### PR DESCRIPTION
**Pull request by jinnoflife**
Added CI-Options to execute the generation with a defined stack
- Possible to set:
  - Stack
  - Offset
  - force (will per default skip already created images)
refactored the execute()-method

Added/Updated PHPDocs

### CODE REVIEW

File: Commands/GenerateWebpImages.php [hunk 1]
```diff
@@ -2,22 +2,87 @@
...
 
 /**
  * Class GenerateWebpImages
  */
 class GenerateWebpImages extends ShopwareCommand
 {
+    /**
+     * @var ModelManager|null
+     */
+    private $modelManager = null;
```
Comment (shyim) [inline Commands/GenerateWebpImages.php]: Can you remove the useless "= null" here? And do a line break here also :)

File: Commands/GenerateWebpImages.php [hunk 2]
```diff
@@ -2,22 +2,87 @@
...
+    private $mediaService = null;
+
+    /**
+     * GenerateWebpImages constructor.
+     *
+     * @param ModelManager                              $manager
+     * @param \Shopware\Bundle\MediaBundle\MediaService $mediaService
+     * @param WebpEncoderFactory                        $webpEncoder
+     */
+    public function __construct(ModelManager $manager, \Shopware\Bundle\MediaBundle\MediaService $mediaService, WebpEncoderFactory $webpEncoder, array $webpConfig)
```
Comment (shyim) [inline Commands/GenerateWebpImages.php]: Import it

File: Commands/GenerateWebpImages.php [hunk 3]
```diff
@@ -2,22 +2,87 @@
...
+    private $runnableEncoders = null;
+
+    /** @var |null */
+    private $webpQuality = null;
+
+    /** @var \Shopware\Bundle\MediaBundle\MediaService|null */
+    private $mediaService = null;
+
+    /**
+     * GenerateWebpImages constructor.
```
Comment (shyim) [inline Commands/GenerateWebpImages.php]: Is useless :)

File: Commands/GenerateWebpImages.php [hunk 4]
```diff
@@ -28,36 +93,70 @@ protected function configure()
...
+
+        $this->buildImageStack($output, $mediaCount, $arguments);
+    }
+
+    /**
+     * @param OutputInterface $output
+     * @param int             $mediaCount
+     * @param Arguments       $arguments
+     */
+    protected function buildImageStack(OutputInterface $output, int $mediaCount, Arguments $arguments)
```
Comment (shyim) [inline Commands/GenerateWebpImages.php]: We have to support PHP 5.6 due Store

File: Commands/GenerateWebpImages.php [hunk 5]
```diff
@@ -28,36 +93,70 @@ protected function configure()
...
-        foreach ($media as $item) {
+    }
+
+    /**
+     * @param bool            $force
+     * @param OutputInterface $output
+     * @param array           $stackMedia
+     * @param ProgressBar     $progress
+     */
+    protected function buildImagesByStack(bool $force, OutputInterface $output, array $stackMedia, ProgressBar $progress)
```
Comment (shyim) [inline Commands/GenerateWebpImages.php]: Remove please bool

File: Components/ImageStack/Arguments.php [hunk 6]
```diff
@@ -0,0 +1,94 @@
+<?php
+/**
+ * Created by PhpStorm.
```
Comment (shyim) [inline Components/ImageStack/Arguments.php]: Can you remove the default phpstorm header?

File: Components/ImageStack/Arguments.php [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+    /**
+     * Arguments constructor.
+     *
+     * @param array $collectionsToUse
+     * @param array $collectionsToIgnore
+     * @param int   $stack
+     * @param int   $offset
+     */
+    public function __construct(array $collectionsToUse, array $collectionsToIgnore, int $stack, int $offset, bool $force)
```
Comment (shyim) [inline Components/ImageStack/Arguments.php]: Can you remove everywhere the types?

File: Factories/WebpConvertFactory.php [hunk 8]
```diff
@@ -0,0 +1,34 @@
+<?php
+/**
+ * Created by PhpStorm.
+ * User: jinnoflife
```
Comment (shyim) [inline Factories/WebpConvertFactory.php]: Header

File: Commands/GenerateWebpImages.php [hunk 9]
```diff
@@ -2,22 +2,87 @@
...
     protected function configure()
     {
         $this
             ->setName('frosh:webp:generate')
-            ->setDescription('Generate webp images for all orginal images');
+            ->setDescription('Generate webp images for all orginal images')
+            ->addOption('stack', 's', InputOption::VALUE_OPTIONAL, 'process amount per iteration')
+            ->addOption('offset', 'o', InputOption::VALUE_OPTIONAL, 'process amount per iteration')
+            ->addOption('force', 'f', InputOption::VALUE_NONE, 'forces recreation')
+            ->addOption('setCollection', 'c', InputOption::VALUE_OPTIONAL | InputOption::VALUE_IS_ARRAY, 'only generates medias for specified collection')
```
Comment (shyim) [inline Commands/GenerateWebpImages.php]: Maybe add to the command description an example option value

**Comment by shyim** [comment]
It looks like it would also resolve the PR #42

File: Commands/GenerateWebpImages.php (hunk 9 above)
Comment (jinnoflife) [reply Commands/GenerateWebpImages.php]: Is this [description](https://github.com/FriendsOfShopware/FroshWebP/blob/59ea13d4cefefb20b2e88f4ca2a5d5d523ad1646/Commands/GenerateWebpImages.php#L69) okay?

**Comment by jinnoflife** [comment]
> It looks like it would also resolve the PR #42 ?

If I understood the PR correctly it would.

File: Commands/DownloadGoogleBinaries.php [hunk 10]
```diff
@@ -48,10 +52,85 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+
+    /**
+     * @return bool
+     */
+    protected function is64bit()
+    {
+        return strpos(php_uname('m'), '64') !== false;
+    }
+
+    /**
```
Comment (shyim) [inline Commands/DownloadGoogleBinaries.php]: wat

File: Commands/DownloadGoogleBinaries.php [hunk 11]
```diff
@@ -48,10 +52,85 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+        }
     }
 
+    /**
+     * @param SymfonyStyle $style
+     * @return int
+     */
+    /**
+     * @param SymfonyStyle $style
+     * @return int
```
Comment (shyim) [inline Commands/DownloadGoogleBinaries.php]: wtf :D

File: Commands/GenerateWebpImages.php [hunk 12]
```diff
@@ -2,22 +2,93 @@
...
     {
         $this
             ->setName('frosh:webp:generate')
-            ->setDescription('Generate webp images for all orginal images');
+            ->setDescription('Generate webp images for all orginal images. Can also run as stack-execution,
+            for specific folders only and with excluded folders')
+            ->addOption('stack', 's', InputOption::VALUE_OPTIONAL, 'process amount per iteration')
+            ->addOption('offset', 'o', InputOption::VALUE_OPTIONAL, 'process amount per iteration')
+            ->addOption('force', 'f', InputOption::VALUE_NONE, 'forces recreation')
+            ->addOption('setCollection', 'c', InputOption::VALUE_OPTIONAL | InputOption::VALUE_IS_ARRAY,
```
Comment (larsbo) [inline Commands/GenerateWebpImages.php]: why not using the common Shopware wording 'albumid'?

Comment (jinnoflife) [reply Commands/GenerateWebpImages.php]: I didn't use it because on my opinion it would occur to much.
But if it is desired, I can change it

**Comment by jinnoflife** [comment]
@shyim Is there anything to do for me?

**Comment by shyim** [comment]
I will look tomorrow 😅

**Pull request merged**